libazureinit/src/goalstate.rs

```diff
@@ -41,34 +45,82 @@ pub struct RoleInstance {
     instance_id: String,
 }
 
-pub async fn get_goalstate(client: &Client) -> Result<Goalstate, Error> {
-    let url = "http://168.63.129.16/machine/?comp=goalstate";
+const DEFAULT_GOALSTATE_URL: &str =
+    "http://168.63.129.16/machine/?comp=goalstate";
+
+pub async fn get_goalstate(
+    client: &Client,
+    retry_interval: Duration,
+    total_timeout: Duration,
+    url: Option<&str>,
+) -> Result<Goalstate, Error> {
+    let url = url.unwrap_or(DEFAULT_GOALSTATE_URL);
 
     let mut headers = HeaderMap::new();
     headers.insert("x-ms-agent-name", HeaderValue::from_static("azure-init"));
     headers.insert("x-ms-version", HeaderValue::from_static("2012-11-30"));
 
-    let request = client.get(url).headers(headers);
-    let response = request.send().await?;
+    let response = timeout(total_timeout, async {
+        loop {
+            let now = std::time::Instant::now();
 
-    if response.status().is_success() {
-        let body = response.text().await?;
+            if let Ok(response) = client
+                .get(url)
+                .headers(headers.clone())
+                .timeout(Duration::from_secs(http::WIRESERVER_HTTP_TIMEOUT_SEC))
+                .send()
+                .await
+            {
+                let statuscode = response.status();
 
-        let goalstate: Goalstate = from_str(&body)?;
-        Ok(goalstate)
-    } else {
-        Err(Error::HttpStatus {
-            endpoint: url.to_owned(),
-            status: response.status(),
-        })
-    }
+                if statuscode == StatusCode::OK {
+                    tracing::info!(
+                        "HTTP response succeeded with status {}",
+                        statuscode
+                    );
+                    return Ok(response);
+                }
+
+                if !http::RETRY_CODES.contains(&statuscode) {
+                    return response.error_for_status().map_err(|error| {
+                        tracing::error!(
+                            ?error,
+                            "{}",
+                            format!(
+                                "HTTP call failed due to status {}",
+                                statuscode
+                            )
+                        );
+                        error
+                    });
+                }
+            }
+
+            _ = total_timeout.saturating_sub(now.elapsed()).as_secs();
+            tracing::info!("Retrying to get HTTP response in {} sec, remaining timeout {} sec.", retry_interval.as_secs(), total_timeout.as_secs());
```

Comment:
Based on my reading of the API, [saturating_sub](https://doc.rust-lang.org/std/time/struct.Duration.html#method.saturating_sub) takes ownership of `total_timeout` and the value returned from `now.elapsed()` and returns a _new_ `Duration` with the result. Since `total_timeout` isn't declared as `mut` that also makes me expect this line to have no side-effects.

I think what you want is
```suggestion
            tracing::info!("Retrying to get HTTP response in {} sec, remaining timeout {} sec.", retry_interval.as_secs(),  total_timeout.saturating_sub(now.elapsed()).as_secs());
```